### PR TITLE
fix: Handle rounding just above kMaxRep more accurately

### PR DESCRIPTION
**Pull request by ximinez**
## High Level Overview of Change

- Treat values in between `kMaxRep` (`2^63-1`) and `kMaxRepUp` (`((kMaxRep
  / 10) + 1) * 10`, which is the next multiple of 10 above `kMaxRep`) as if
  those values were sequential, and values in between were "fractional".
- This results in values above the midpoint rounding up to `kMaxRepUp`,
  and below the midpoint to `kMaxRep` when rounding to nearest. Other
  rounding modes act along the same lines.

### CODE REVIEW

File: src/test/basics/Number_test.cpp [hunk 1]
```diff
@@ -2146,6 +2265,83 @@ class Number_test : public beast::unit_test::Suite
...
+            // Both ToNearest and Downward should round to `below`
+            auto const actual = static_cast<std::uint64_t>(kMaxRep) + 1;
+            Number const below{static_cast<std::int64_t>(kMaxRep), 0};
+            Number const above{
+                false, static_cast<std::uint64_t>(kMaxRep) + 3, 0, Number::Normalized{}};
+
+            auto construct = [](Number::RoundingMode mode) {
+                NumberRoundModeGuard const roundGuard{mode};
+                return Number(false, actual, 0, Number::Normalized{});
+            };
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/test/basics/Number_test.cpp]: > The `construct` lambda uses `actual` but has an empty capture list, which will not compile ("actual" is not captured). Capture `actual` by value (or reference) so the Number can be constructed under the selected rounding mode.

_"E pur si muove"_

`actual` is implicitly captured, and compiles just fine. In fact, clang warns if `actual` is explicitly captured.

File: src/libxrpl/basics/Number.cpp [hunk 2]
```diff
@@ -414,10 +418,48 @@ Number::Guard::doDropDigit<uint128_t>(uint128_t& mantissa, int& exponent) noexce
...
+        //      * it can never be exactly at the midpoint, because kMaxRepUp is always even, and
+        //        kMaxRep is always odd, so don't worry about that case.
+        // * For round upward, will round up to kMaxRepUp for positive values, down to kMaxRep for
+        //   negative.
+        // * For round downward, does the opposite of upward.
+        // * For round toward zero, always rounds down to kMaxRep.
+        auto constexpr spread = kMaxRepUp - kMaxRep;
+        static_assert(spread < 10 && spread > 0);
+        // This should absolutely be impossible
+        if constexpr (spread == 0)
```
Comment (TimothyBanks) [inline src/libxrpl/basics/Number.cpp]: I am a bit uncertain why this is needed.  Both the `static_assert` and `if constexpr` would be evaluated at compile time.  Am I overlooking a runtime path that could get here?

**Review by TimothyBanks — DISMISSED** [review]

File: src/libxrpl/basics/Number.cpp [hunk 3]
```diff
@@ -414,10 +418,48 @@ Number::Guard::doDropDigit<uint128_t>(uint128_t& mantissa, int& exponent) noexce
...
+        //      * if the mantissa is below the midpoint, it'll round "down" to kMaxRep
+        //      * if above the midpoint, it'll round "up" to kMaxRepUp
+        //      * it can never be exactly at the midpoint, because kMaxRepUp is always even, and
+        //        kMaxRep is always odd, so don't worry about that case.
+        // * For round upward, will round up to kMaxRepUp for positive values, down to kMaxRep for
+        //   negative.
+        // * For round downward, does the opposite of upward.
+        // * For round toward zero, always rounds down to kMaxRep.
+        auto constexpr spread = kMaxRepUp - kMaxRep;
+        static_assert(spread < 10 && spread > 0);
```
Comment (gregtatcam) [inline src/libxrpl/basics/Number.cpp]: Why not make it an exact assert? 

    static_assert(spread == 3);

Sure, the range is future proof but the logic is not and might not work for other spreads. Making the spread exact is a red flag when the spread changes.

File: src/libxrpl/basics/Number.cpp [hunk 4]
```diff
@@ -586,6 +651,14 @@ Number::Guard::doRound(rep& drops, std::string location) const
         }
         ++drops;
     }
+    else if (
```
Comment (gregtatcam) [inline src/libxrpl/basics/Number.cpp]: This condition is never reachable: `kMaxRep = std::numeric_limits<rep>::max();` so there is no way `drops`, which is `rep` could be `kMaxRepUp`. So this clamp to `kMaxRep` doesn't work. Maybe just have an assert. Maybe instead of this `else if` add static assert 
   
    static_assert(Number::kMaxRep == std::numeric_limits<rep>::max());

or XRPL_ASSERT since doRound() is called by rep(), which converts negative to positive. 

    XRPL_ASSERT(drops >= 0, "xrpl::Number::Guard::doRound : positive magnitude");

or maybe add both.

**Review by gregtatcam — CHANGES_REQUESTED** [review]
`pushOverflow()` handles the simple integer cases, but it loses information when there are already guard digits. For example, `kMaxRep + 1.6` should round up because it is closer to `kMaxRepUp` than to `kMaxRep`. Today the code sees the integer part, `kMaxRep + 1`, turns that into a fake guard digit 3, and pushes it in front of the real guard digit 6. That makes the value look like .36 of the way through the gap instead of 1.6 / 3, so `ToNearest` rounds down incorrectly. I think the fix is to make the cusp rounding decision use the integer offset and the existing guard digits together, instead of adding a fake digit on top of the current guard state. Here is the unit test, which demonstrates the issue:

            NumberRoundModeGuard const roundGuard{Number::RoundingMode::ToNearest};

            Number const below{static_cast<std::int64_t>(Number::kMaxRep), 0};
            Number const above{false, Number::kMaxRepUp, 0, Number::Normalized{}};

            // Exact value is kMaxRep + 1.6. Since the gap from kMaxRep to kMaxRepUp is 3,
            // this is closer to kMaxRepUp than kMaxRep.
            Number const actual = below + Number{16, -1};

            std::stringstream ss;
            ss << "kMaxRep + 1.6 rounded to " << actual << ". Expected: " << above;
            BEAST_EXPECTS(actual == above, ss.str()); // fails

File: src/libxrpl/basics/Number.cpp (hunk 2 above)
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp]: > I am a bit uncertain why this is needed. Both the `static_assert` and `if constexpr` would be evaluated at compile time. Am I overlooking a runtime path that could get here?

I think I was just being overly paranoid. I'll remove the extra `if constexpr`.

File: src/libxrpl/basics/Number.cpp (hunk 3 above)
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp]: > Why not make it an exact assert?
> 
> ```
> static_assert(spread == 3);
> ```
> 
> Sure, the range is future proof but the logic is not and might not work for other spreads. Making the spread exact is a red flag when the spread changes.

My intention was that the logic _would_ be future-proof. Also, it turns out that `2^(2^x - 1) - 1` where `x > 1` always ends in 7, so the spread is going to be 3 for any size mantissa that we're likely to use.

I'll make the change.

File: src/libxrpl/basics/Number.cpp (hunk 4 above)
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp]: > This condition is never reachable: `kMaxRep = std::numeric_limits<rep>::max();` so there is no way `drops`, which is `rep` could be `kMaxRepUp`. So this clamp to `kMaxRep` doesn't work. Maybe just have an assert. Maybe instead of this `else if` add static assert

Yeah, I thought so! Again, I was being paranoid. I didn't think about the simple fact that `drops` is `rep`.

**Comment by ximinez** [comment]
> `pushOverflow()` handles the simple integer cases, but it loses information when there are already guard digits. For example, `kMaxRep + 1.6` should round up because it is closer to `kMaxRepUp` than to `kMaxRep`. Today the code sees the integer part, `kMaxRep + 1`, turns that into a fake guard digit 3, and pushes it in front of the real guard digit 6. That makes the value look like .36 of the way through the gap instead of 1.6 / 3, so `ToNearest` rounds down incorrectly. I think the fix is to make the cusp rounding decision use the integer offset and the existing guard digits together, instead of adding a fake digit on top of the current guard state. Here is the unit test, which demonstrates the issue:

Oh, daaaaaaannnnnngggggg... Good catch.

File: src/libxrpl/basics/Number.cpp [hunk 5]
```diff
@@ -455,16 +525,19 @@ Number::Guard::round() const noexcept
...
     if (mantissa < minMantissa)
     {
         mantissa *= 10;
         --exponent;
     }
-    if (exponent < kMinExponent)
+    // mantissa should never be 0, but if it _is_ make the result kZero.
+    if (exponent < kMinExponent ||
+        (cuspRoundingFix >= MantissaRange::CuspRoundingFix::Enabled330 && mantissa == 0))
     {
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp:525]: > `bringIntoRange` asserts `mantissa != 0`, but then immediately has a code path that tries to handle `mantissa == 0` by forcing the result to `kZero`. In debug builds this makes the `mantissa == 0` handling unreachable and can turn a recoverable edge case into an assertion failure. Either remove the `mantissa == 0` branch or handle it before asserting.

Yeah, it's a fallback recovery mechanism for release builds so we don't end up with an inconsistent / invalid state.

I'll shuffle some stuff around to make that clearer.

File: src/test/app/LoanBroker_test.cpp [hunk 6]
```diff
@@ -1452,18 +1452,54 @@ class LoanBroker_test : public beast::unit_test::Suite
...
         }
 
+        if (Number::getMantissaScale() == MantissaRange::MantissaScale::Large330)
         {
-            auto const dm = power(2, 63) - 1;
-            BEAST_EXPECTS(dm > kMaxMpTokenAmount, to_string(dm));
-            tx2[sfDebtMaximum] = dm;
-            env(tx2, Ter(temINVALID));
-        }
+            // For the Large330 scale, 2^63 rounds _down_ to Number::kMaxRep
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/test/app/LoanBroker_test.cpp]: Will fix

File: src/test/app/LoanBroker_test.cpp [hunk 7]
```diff
@@ -1452,18 +1452,54 @@ class LoanBroker_test : public beast::unit_test::Suite
             env(tx2, Ter(temINVALID));
         }
 
+        if (Number::getMantissaScale() >= MantissaRange::MantissaScale::Large330)
```
Comment (gregtatcam) [inline src/test/app/LoanBroker_test.cpp:1457]: Consider adding this test above line 1455:

        {
            // Raw input one unit above the maximum MPT amount must not be
            // accepted after Number normalization.
            tx2[sfDebtMaximum] = "9223372036854775808";
            env(tx2, Ter(temINVALID));
        }

File: src/test/app/Vault_test.cpp [hunk 8]
```diff
@@ -5606,9 +5606,9 @@ class Vault_test : public beast::unit_test::Suite
         BEAST_EXPECT(maxInt64 == "9223372036854775807");
 
         // Naming things is hard
-        auto const maxInt64Plus1 = std::to_string(
-            static_cast<std::uint64_t>(std::numeric_limits<std::int64_t>::max()) + 1);
-        BEAST_EXPECT(maxInt64Plus1 == "9223372036854775808");
+        auto const maxInt64Plus2 = std::to_string(
```
Comment (gregtatcam) [inline src/test/app/Vault_test.cpp:5255]: Consider adding this above and then a test below. I have another comment below.

            auto const maxInt64Plus1 = std::to_string(
            static_cast<std::uint64_t>(std::numeric_limits<std::int64_t>::max()) + 1);
        BEAST_EXPECT(maxInt64Plus1 == "9223372036854775808");

File: src/test/app/Vault_test.cpp [hunk 9]
```diff
@@ -5636,15 +5636,15 @@ class Vault_test : public beast::unit_test::Suite
             env(tx);
             env.close();
 
-            tx[sfAssetsMaximum] = maxInt64Plus1;
+            tx[sfAssetsMaximum] = maxInt64Plus2;
             env(tx, Ter(tefEXCEPTION));
             env.close();
```
Comment (gregtatcam) [inline src/test/app/Vault_test.cpp:5315]: Consider adding this test:

            tx[sfAssetsMaximum] = maxInt64Plus1;
            env(tx, Ter(tefEXCEPTION));
            env.close();

**Review by gregtatcam — DISMISSED** [review]
LGTM

Could we add explicit boundary coverage for raw 2^63 input?
Specifically:
LoanBrokerSet: set sfDebtMaximum to "9223372036854775808" and expect temINVALID.
VaultCreate: for MPT sfAssetsMaximum, set it to maxInt64 + 1 / "9223372036854775808" and expect failure.
The existing tests cover nearby rounded Number cases and maxInt64 + 2, but this exact raw-input boundary is worth locking down because it verifies invalid MPT-sized input is not normalized down and accepted.

I added comments for the suggested tests to LoanBroker_test, Vault_test.

File: src/test/app/LoanBroker_test.cpp (hunk 7 above)
Comment (ximinez) [reply src/test/app/LoanBroker_test.cpp:1457]: > Consider adding this test above line 1455:
> 
> ```
>     {
>         // Raw input one unit above the maximum MPT amount must not be
>         // accepted after Number normalization.
>         tx2[sfDebtMaximum] = "9223372036854775808";
>         env(tx2, Ter(temINVALID));
>     }
> ```

But that's not the expected behavior. With `ToNearest` rounding, which is the default, the value '9...808' is going to round to '9...807' as soon as that Json string is parsed into a `Number`. It would be roughly the equivalent to the value being "100.25". It's going to round to 100 as soon as it becomes an integer. The difference is that `9...808` _can't_ be represented as a `Number`. It _has_ to be rounded one way or the other _before_ it's serialized into a transaction, whereas "100.25" can be, so it'll get through to the tx engine as-is, and so for an MPT, will return tecPRECISION_LOSS.

Actually, a better comparison might be if the value was `9...807.1`.

There is another option, that might have the effect you're expecting. The problem here is in the parsing of Json to a `Number`, so let's add some kind of check there. Specifically `numberToJson`.

Yes, I'll add checks that the two representations (pre-normalization `NumberParts`, and normalized `Number`) are the same, and throw if they don't match.

**Review by gregtatcam — CHANGES_REQUESTED** [review]
STNumber.cpp (line 266) now asserts that the normalized Number sign matches the parsed sign. That breaks canonical zero inputs like "-0" because parsing preserves parts.negative=true, while Number normalizes zero to non-negative. Existing tests at STNumber_test.cpp (line 93) still expect these to parse as zero:

xrpl.protocol.STNumber Small
Assertion failed: (("numberFromJson : signs agree") && (negative == parts.negative)), function numberFromJson, file STNumber.cpp, line 266.
zsh: abort      ./xrpld -u STNumber


---------------------

Number_test.cpp (line 1430) has the expected string backwards. Number(-2, 11) is -200000000000, so adding 1 should produce -199999999999, not -200000000001

File: src/test/app/Vault_test.cpp [hunk 10]
```diff
@@ -5636,25 +5641,57 @@ class Vault_test : public beast::unit_test::Suite
...
+                env(tx);
+                // should throw in parser
+                fail();
+            }
+            catch (std::exception const& e)
+            {
+                BEAST_EXPECT(
+                    std::string(e.what()) ==
+                    "invalidParamsField 'tx_json.AssetsMaximum' has invalid data.");
+            }
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/test/app/Vault_test.cpp:5334]: > `env.setParseFailureExpected(true)` is not reset to `false` at the end of this block. `parseFailureExpected_` suppresses parse-failure logging in `Env::autofill` (src/test/jtx/impl/Env.cpp:604), so leaving it enabled can hide unexpected parse errors in subsequent test steps.

That's intentional in this case. There are several more parse failures in the rest of this function, so I set the flag once for all of them. Notably, even if the logging is skipped, that function still throws, so the important thing - a test failure - still happens.

File: src/test/app/LoanBroker_test.cpp [hunk 11]
```diff
@@ -1452,18 +1452,76 @@ class LoanBroker_test : public beast::unit_test::Suite
...
             env(tx2, Ter(temINVALID));
+            // should throw in parser
+            fail();
+        }
+        catch (std::exception const& e)
+        {
+            BEAST_EXPECT(
+                std::string(e.what()) ==
+                "invalidParamsField 'tx_json.DebtMaximum' has invalid data.");
         }
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/test/app/LoanBroker_test.cpp:1456]: > `env.setParseFailureExpected(true)` is not reset to `false` after the expected parse failure. `parseFailureExpected_` only suppresses parse-failure logging (src/test/jtx/impl/Env.cpp:604), so leaving it enabled can hide unexpected parse errors later in this test.

Since this is the only test case expected to have a parsing problem, I will re-disable the flag for this one.

**Review by gregtatcam — CHANGES_REQUESTED** [review]
STNumber.cpp (line 281)] rejects one exact boundary value. With minMantissa = Number::kMaxRep / 10, the loop uses <, so parsed "922337203685477580" is not scaled to 9223372036854775800e-1. But Number does normalize it that way because 922337203685477580 * 10 still fits within kMaxRep. The final comparison then mismatches and throws "number can not be represented" even though the value is exactly representable under the large mantissa scales. The same applies to the negative value. This likely needs <= here, plus a test for that exact boundary. Here is a unit-test that shows the failure:

            {
                auto const parseBoundary = [](std::string const& boundary) {
                    return numberFromJson(sfNumber, boundary);
                };
                auto const expectBoundaryThrows = [this,
                                                   &parseBoundary](std::string const& boundary) {
                    try
                    {
                        auto const bad = parseBoundary(boundary);
                        fail();
                    }
                    catch (std::exception const& e)
                    {
                        BEAST_EXPECT(std::string(e.what()) == "number can not be represented");
                    }
                };

                // Small rejects this; large scales parse it as 9223372036854775800e-1.
                auto constexpr positiveBoundary = "922337203685477580";
                auto constexpr negativeBoundary = "-922337203685477580";
                if (Number::getMantissaScale() == MantissaRange::MantissaScale::Small)
                {
                    expectBoundaryThrows(positiveBoundary);
                    expectBoundaryThrows(negativeBoundary);
                }
                else
                {
                    BEAST_EXPECT(
                        parseBoundary(positiveBoundary) ==
                        STNumber(sfNumber, Number{922'337'203'685'477'580, 0}));
                    BEAST_EXPECT(
                        parseBoundary(negativeBoundary) ==
                        STNumber(sfNumber, Number{-922'337'203'685'477'580, 0}));
                }
            }

**Review by gregtatcam — COMMENTED** [review]
How about this alternative to the exactness validation? It seems simpler:

    auto canonicalize = [](NumberParts p) {
        if (p.mantissa == 0)
            return NumberParts{};

        while (p.mantissa % 10 == 0)
        {
            p.mantissa /= 10;
            ++p.exponent;
        }

        return p;
    };

    auto const numberMantissa = num.mantissa();
    auto const parsed = canonicalize(parts);
    auto const represented = canonicalize(NumberParts{
        .mantissa = Number::externalToInternal(numberMantissa),
        .exponent = num.exponent(),
        .negative = numberMantissa < 0,
    });

    if (
        parsed.mantissa != represented.mantissa ||
        parsed.exponent != represented.exponent ||
        parsed.negative != represented.negative)
    {
        Throw<std::runtime_error>("number can not be represented");
    }

**Comment by ximinez** [comment]
> How about this alternative to the exactness validation? It seems simpler:

I thought it might be moderately faster to do a comparison and multiply than to do a modulo and a divide. It would also take fewer steps if both values started with their exponents close to each other. And certainly faster if the exponents started equivalent.

But this seems a lot simpler and clearer about what it's doing, so yes, I'll steal it, but with some tweaks.

**Review by gregtatcam — DISMISSED** [review]
LGTM

**Review by gregtatcam — COMMENTED** [review]
There appears to be a boundary issue when normalizing `kMaxRepUp` at `kMaxExponent`.

`kMaxExponent` itself is valid. However, `kMaxRepUp` is larger than the public `Number::mantissa()` range, so the public accessors rescale it by returning `mantissa() == kMaxRepUp / 10` and `exponent() == kMaxExponent + 1`.

For example:

Number{false, Number::kMaxRepUp, Number::kMaxExponent, Number::Normalized{}}

is accepted, but its external view is:

mantissa = 922337203685477581
exponent = 32769

That external exponent is outside the documented `[kMinExponent, kMaxExponent]` range, and the represented value is greater than `Number::max()`. I think this should overflow during normalization rather than producing an externally out-of-range `Number`.

This unit-test in testLimits() confirms the issue:

        if (scale >= MantissaRange::MantissaScale::Large330)
        {
            caught = false;
            std::int64_t mantissa = 0;
            int exponent = 0;
            std::string text;
            try
            {
                Number const q{
                    false, Number::kMaxRepUp, Number::kMaxExponent, Number::Normalized{}};
                mantissa = q.mantissa();
                exponent = q.exponent();
                text = to_string(q);
            }
            catch (std::overflow_error const&)
            {
                caught = true;
            }
            std::stringstream ss;
            ss << "Expected kMaxRepUp at kMaxExponent to overflow. Got " << text << " (mantissa "
               << mantissa << ", exponent " << exponent << ").";
            BEAST_EXPECTS(caught, ss.str());
        }

**Comment by ximinez** [comment]
> `kMaxExponent` itself is valid. However, `kMaxRepUp` is larger than the public `Number::mantissa()` range, so the public accessors rescale it by returning `mantissa() == kMaxRepUp / 10` and `exponent() == kMaxExponent + 1`.
> 
> For example:
> 
> Number{false, Number::kMaxRepUp, Number::kMaxExponent, Number::Normalized{}}
> 
> is accepted, but its external view is:
> 
> mantissa = 922337203685477581 exponent = 32769

Good news!

```
    // The range for the exponent when normalized
    static constexpr int kMinExponent = -32768;
    static constexpr int kMaxExponent = 32768;
```

The max exponent is only relevant in the normalized / internal view. That example external view is no longer normalized.

Secondarily, the exponent range is intentionally extreme - well beyond any value that would reasonably exist related to the ledger. `STAmount` is something like -80 to 96. It would take some awfully big math to get from there to 32,768. Because of that, I'm not really worried about what happens right around the edges, as long as it's consistent.

I've added some unit tests, though.

**Review by gregtatcam — DISMISSED** [review]

**Review by TimothyBanks — DISMISSED** [review]

File: src/libxrpl/basics/Number.cpp [hunk 12]
```diff
@@ -466,7 +542,9 @@ template <UnsignedMantissa T>
 void
 Number::Guard::doRoundUp(bool& negative, T& mantissa, int& exponent, std::string location)
 {
-    auto r = round();
+    pushOverflow(mantissa);
+
+    auto const r = round();
     if (r == Round::Up || (r == Round::Even && (mantissa & 1) == 1))
     {
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp:545]: > With cusp rounding Enabled330, an `Even` guard (exactly 0.5 in the dropped digit) at `mantissa == kMaxRep` should not trigger the usual tie-to-even increment. At this boundary the next *representable* value is `kMaxRepUp` (not `kMaxRep + 1`), so treating `0.5` as a tie can incorrectly drive recursion / scaling and potentially round up when it should round down to `kMaxRep`. Consider overriding `r == Round::Even` to `Round::Down` in this specific cusp case before applying the tie-to-even rule.

Yeah, that's exactly what `pushOverflow` does.

**Review by TimothyBanks — DISMISSED** [review]

**Review by gregtatcam — DISMISSED** [review]

File: src/libxrpl/basics/Number.cpp [hunk 13]
```diff
@@ -394,10 +398,70 @@ Number::Guard::doDropDigit<uint128_t>(uint128_t& mantissa, int& exponent) noexce
...
+        //        kMaxRep is always odd, so don't worry about that case.
+        // * For round upward, will round up to kMaxRepUp for positive values, down to kMaxRep for
+        //   negative.
+        // * For round downward, does the opposite of upward.
+        // * For round toward zero, always rounds down to kMaxRep.
+
+        auto const diff = mantissa - kMaxRep;
+        auto const digit = (diff * 10) / spread;
+        XRPL_ASSERT(
+            digit >= 0 && digit < 10 && digit != 5,
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp]: Will fix

**Review by gregtatcam — APPROVED** [review]

**Review by TimothyBanks — APPROVED** [review]
